snekbox/memfs.py

```diff
@@ -0,0 +1,148 @@
+"""Memory filesystem for snekbox."""
+from __future__ import annotations
+
+import logging
+from collections.abc import Generator
+from functools import cached_property
+from pathlib import Path
+from types import TracebackType
+from typing import Type
+from uuid import uuid4
+
+from snekbox.libmount import mount, unmount
+from snekbox.snekio import FileAttachment
+
+log = logging.getLogger(__name__)
+
+__all__ = ("MemFS", "parse_files")
+
+
+def parse_files(
+    fs: MemFS,
+    files_limit: int,
+    files_pattern: str,
+    preload_json: bool = False,
+) -> list[FileAttachment]:
+    """
+    Parse files in a MemFS.
+
+    Args:
+        fs: The MemFS to parse.
+        files_limit: The maximum number of files to parse.
+        files_pattern: The glob pattern to match files against.
+        preload_json: Whether to preload and cache JSON data.
+
+    Returns:
+        List of FileAttachments sorted lexically by path name.
+    """
+    res = sorted(fs.attachments(files_limit, files_pattern), key=lambda file: file.path)
+    if preload_json:
+        for file in res:
+            _ = file.json
+    return res
+
+
+class MemFS:
+    """A temporary directory using tmpfs."""
+
+    def __init__(self, instance_size: int, root_dir: str | Path = "/memfs") -> None:
+        """
+        Create a temporary directory using tmpfs.
+
+        Args:
+            instance_size: Size limit of each tmpfs instance in bytes.
+            root_dir: Root directory to mount instances in.
+        """
+        self.instance_size = instance_size
+        self._path: Path | None = None
+        self.root_dir: Path = Path(root_dir)
+        self.root_dir.mkdir(exist_ok=True, parents=True)
+
+    @cached_property
+    def path(self) -> Path:
+        """Returns the path of the MemFS."""
+        if self._path is None:
+            raise RuntimeError("MemFS accessed before __enter__.")
```

Comment:
Sorry to sort of "go back" on this, but it dawned on me: why is this being set in `__enter__` instead of `__init__`? Most libraries I've seen support using an object without a context manager (requiring a manual call to some public cleanup function). Wouldn't it make more sense for all of `__enter__` to just be in `__init__` instead? This current situation of being able to instantiate an object in an invalid state just feels really wrong. On the other hand, having side effects when instantiation an object also doesn't sound great, but the former seems like the worse of the two.

I don't want to block the PR on this. This is all internal stuff we can revisit at a later time if you prefer.

Reply:
Yeah I agree, I think the context manager being optional is more in-line with conventions. I've changed it to be initialized during `__init__`, with a `weakref.finalize` running the cleanup on garbage collection with a `ResourceWarning` if the context manager is not used, modeled after https://github.com/python/cpython/blob/main/Lib/tempfile.py#L844